docs/source/design/user_messaging.rst

```diff
@@ -0,0 +1,124 @@
+and are used to inform generic error handling, as well as particular calling commands
+on how to proceed with a specific operation.
+
+Yield result records when:
+
+1. A generic and standard way of error handling is considered useful
+2. A process can reasonably be continued despite evident errors or shortcomings in
+   its subprocesses
+
+Based on the ``status`` field of a result record, a result is categorized into
+*success* (``ok``, ``notneeded``) and *failure* (``impossible``, ``error``).
+Depending on the ``on_failure`` parameterization of a command call (``on_failure='stop'``,
+``on_failure='continue'``, or ``on_failure='ignore'``), any failure-result
+emitted by a command can lead to an ``IncompleteResultsError`` being raised on command
+exit, or a non-zero exit code on the command line.
+
+
+Exception handling
+------------------
+
+In general, **exceptions should be raised when there is no way to ignore or recover from
+the offending action**.
+
+More specifically, raise an exception when:
+
+1. A command's parameter specifications are violated
+2. An additional requirement (beyond parameters) for the successful running of a
+   command, function, or process is not met
+
+It must be made clear to the user/caller what the exact cause of the exception
+is, given the context within which the user/caller triggered the action.
+This is achieved directly via a (re)raised exception, as opposed to logging messages or
+results records which could be ignored or unseen by the user.
+
+If the relevant caller receives a set of result records, these should be inspected
+individually in order to identify result status values that could require an exception
+to be raised to the user/caller. Depending on developer-defined logic, an exception can
+then be raised with an unambiguous message that excludes internal and intermediate
+result messages.
+
+Additionally, in the case of a complex set of dependent actions it could be expensive to
+confirm parameter violations. In such cases, initial sub-processes might already generate
+result records that have to be inspected by the caller, and it could be practically better
+to yield a result record (with ``status=[error|impossible]``) to communicate the failure.
+It would then be up to the upstream caller to decide whether to specify
+``on_failure='ignore'`` or whether to inspect individual result records and turn them
+into exceptions or not.
+
+
+Logging
+-------
+
+Logging provides developers with additional means to describe steps in a process,
+so as to **allow insight into the program flow during debugging** or analysis of e.g.
+usage patterns. Logging can be turned off externally, filtered, and redirected. Apart from
+the log-level and message, it is not inspectable and cannot be used to control the logic
+or flow of a program.
+
+Importantly, logging is not a user messaging method. Therefore:
+
+1. No command should rely on logging for user communication.
```

Comment:
ouch, there are many commands that are guilty of that. 
I wonder whether the INFO messages that are emitted via logging could be an exception to this rule - they couldn't be communicated via an exception or a result record (because they are neither - they're in-progress communication, such as progress bars.

Reply:
Ok, makes sense to me to shift the focus to:

- logging should not be the primary a user messaging method for command outcomes
- use logging for in-progress user communication (e.g. info messages or progress bars)